Add Makefile and update Makefile.uk

eduardvintila opened the pull request:
This PR is part of a series of PRs that work in conjunction for the update of Go support to 1.18:

- https://github.com/unikraft/lib-libgo/pull/7
- https://github.com/unikraft/unikraft/pull/1005
- https://github.com/unikraft/lib-musl/pull/67
- https://github.com/unikraft/app-helloworld-go/pull/9
- https://github.com/unikraft/lib-compiler-rt/pull/16
- https://github.com/unikraft/lib-musl/pull/52

For testing, make sure to pass to QEMU the `-cpu host` argument (this is necessary because the paging API and ukvmem are required by `libgo` as part of Virtual Address space management, and somehow the default CPU just won't handle 1GB pages).

Only GCC12 and x86 is supported for the moment.

RaduNichita commented on Makefile:
```diff
@@ -0,0 +1,15 @@
+UK_ROOT ?= $(PWD)/.unikraft/unikraft
+UK_LIBS ?= $(PWD)/.unikraft/libs
+
+LIBS-y :=
```
Why did you put the libraries like so and not on the same line?

eduardvintila replied:
No particular reason, it was easier to remove and add libraries that way 😆

RaduNichita commented on Makefile.uk:
```diff
@@ -1,3 +1,3 @@
-$(eval $(call addlib,apphelloworldgo))
+$(eval $(call addgoapp,apphelloworldgo))
```
Why do you have this change here?

eduardvintila replied:
Because `libgo` has custom build mechanics, given by [this PR from Unikraft core](https://github.com/unikraft/unikraft/pull/1005/commits/84ecf07cb866fdb12b9ebd6529bb5d5a11fe9e67) and [this rule from `libgo`](https://github.com/unikraft/lib-libgo/blob/79a26b78464635325656dbadf08ef3ff26e30791/Makefile.rules#L164)

RaduNichita reviewed (commented):
It works on my side, many thanks @eduardvintila :godmode: 

I left two questions more out of curiosity, I will approve after receiving your input of those.

StefanJum reviewed (changes requested):
@eduardvintila looks good, few comments:

* Use `LIBS` instead of `LIBS-y`, in order to be consistent with other applications.
* State why changing the `addlib` rule to `addgoapp` was necessary in the commit description.

eduardvintila committed:
Makefile.uk: Change `addlib` to `addgoapp`

Golang applications which use the `libgo` external library have
custom build mechanics, where the `addlib` rule is replaced by
either `addgolib` (for GO packages used as libraries), or `addgoapp`.
In this case, we need to call the `addgoapp` Makefile rule to
indicate that we're building a GO app which provides the main function.

Signed-off-by: Eduard Vintilă <eduard.vintila47@gmail.com>

eduardvintila committed:
Makefile: Introduce Makefile

Signed-off-by: Eduard Vintilă <eduard.vintila47@gmail.com>

eduardvintila commented:
@StefanJum I've added an explanation and replaced `LIBS-y` with `LIBS`. @RaduNichita, I've left the libraries only on one line :)

RaduNichita reviewed (approved):
All good with me now.

Reviewed-by: Radu Nichita <radunichita99@gmail.com>

StefanJum reviewed (approved):
All good, thanks.

Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
